vpr/src/route/route_tree_type.h

```diff
@@ -1,60 +1,210 @@
 
-/* Structure describing one node in a routing tree (used to get net delays   *
- * incrementally during routing, as pieces are being added).                 *
- * u.child_list:  Pointer to a linked list of linked_rt_edge.  Each one of   *
- *                the linked list entries gives a child of this node.        *
- * u.next:  Used only when this node is on the free list.  Gives the next    *
- *          node on the free list.                                           *
- * parent_node:  Pointer to the rt_node that is this node's parent (used to  *
- *               make bottom to top traversals).                             *
- * re_expand:  Should this node be put on the heap as part of the partial    *
- *             routing to act as a source for subsequent connections?        *
- * parent_switch:  Index of the switch type driving this node (by its        *
- *                 parent).                                                  *
- * inode:  index (ID) of the rr_node that corresponds to this rt_node.       *
- * net_pin_index:  Net pin index associated with the rt_node. This value
- *                 ranges from 1 to fanout [1..num_pins-1]. For cases when   *
- *                 different speed paths are taken to the same SINK for      *
- *                 different pins, inode cannot uniquely identify each SINK, *
- *                 so the net pin index guarantees an unique identification  *
- *                 for each SINK rt_node. For non-SINK nodes and for SINK    *
- *                 nodes with no associated net pin index, (i.e. special     *
- *                 SINKs like the source of a clock tree which do not        *
- *                 correspond to an actual netlist connection), the value    *
- *                 for this member should be set to OPEN (-1).               *
- * C_downstream:  Total downstream capacitance from this rt_node.  That is,  *
- *                the total C of the subtree rooted at the current node,     *
- *                including the C of the current node.                       *
- * R_upstream:  Total upstream resistance from this rt_node to the net       *
- *              source, including any device_ctx.rr_nodes[].R of this node.  *
- * Tdel:  Time delay for the signal to get from the net source to this node. *
- *        Includes the time to go through this node.                         */
-struct t_rt_node {
-    union {
-        t_linked_rt_edge* child_list;
-        t_rt_node* next;
-    } u;
-    t_rt_node* parent_node;
-    short parent_switch;
+    RouteTreeNode(RRNodeId inode, RRSwitchId parent_switch, vtr::optional<RouteTreeNode&> parent, vtr::optional<RouteTree&> root);
+
+    RRNodeId inode;
+    RRSwitchId parent_switch;
+    vtr::optional<RouteTreeNode&> parent;
+    vtr::optional<RouteTree&> root;
     bool re_expand;
-    int inode;
     int net_pin_index;
     float C_downstream;
     float R_upstream;
     float Tdel;
+    std::list<RouteTreeNode> _child_nodes;
```

Comment:
Just out of curiosity, the way you define child_node implies that the lifetime of the child nodes is dependent on their parent. In the previous approach, we stored a pointer to the child, so changing a parent of a sub-tree could be done more efficiently. However, with the current approach, wouldn't changing the parent of a sub-tree take longer, since we now have to update the parent of each child node individually?

Reply:
We could optimize for that if that was the case, but current route tree operations only add/delete child nodes from existing nodes and don't do any "re-organization" of the tree. (unlikely in the future too, since rt nodes are more or less "enabled" nodes from the rr graph which stays constant)

> Why not define the fields as private and use setter/getter functions instead?

For `_child_nodes`, gcc complains about the returned reference not being const when it's private. I can try to fix it.

For other fields: Right now route_tree functions live outside the class. This is an artifact from the original code. If I make the fields private and provide getters, I need to move the functions in `route_tree_timing.cpp` into `RouteTreeNode` or `RouteTree` which means another big refactor. If I provide getters and setters, well, that just makes the code longer for the same degree of encapsulation. At least in my opinion.

Reply:
I agree that at some point, the function under route_tree_timing should also be a part of this class. However, I think it would be better if we approach the refactoring in parts. For instance, wrapping the variables under a function would make future refactoring much easier.

Reply:
Also, the comments above this class need to be updated to include the description of _child_nodes

Reply:
It's private now. also most `RouteTreeNode` references are const (only way to get a non-const reference is to create a node), so data fields can stay public